control_plane/attachment_service/src/service.rs

```diff
@@ -3124,11 +3122,48 @@ impl Service {
             old_shard_count,
             new_shard_count,
             new_stripe_size,
-            targets,
+            mut targets,
             policy,
             shard_ident,
         } = params;
 
+        // Drop any secondary locations: pageservers do not support splitting these, and in any case the
+        // end-state for a split tenant will usually be to have secondary locations on different nodes.
+        // The reconciliation calls in this block also implicitly cancel+barrier wrt any ongoing reconciliation
+        // at the time of split.
+        let waiters = {
+            let mut locked = self.inner.write().unwrap();
+            let mut waiters = Vec::new();
+            let (nodes, tenants, scheduler) = locked.parts_mut();
+            for target in &mut targets {
+                let Some(shard) = tenants.get_mut(&target.parent_id) else {
+                    // Paranoia check: this shouldn't happen: we have the oplock for this tenant ID.
+                    return Err(ApiError::InternalServerError(anyhow::anyhow!(
+                        "Shard {} not found",
+                        target.parent_id
+                    )));
+                };
+
+                if shard.intent.get_attached() != &Some(target.node.get_id()) {
+                    // Paranoia check: this shouldn't happen: we have the oplock for this tenant ID.
+                    return Err(ApiError::Conflict(format!(
+                        "Shard {} unexpectedly rescheduled during split",
+                        target.parent_id
+                    )));
+                }
+
+                // Irrespective of PlacementPolicy, clear secondary locations from intent
+                shard.intent.clear_secondary(scheduler);
+
+                // Run Reconciler to execute detach fo secondary locations.
+                if let Some(waiter) = self.maybe_reconcile_shard(shard, nodes) {
+                    waiters.push(waiter);
+                }
+            }
+            waiters
+        };
+        self.await_waiters(waiters, RECONCILE_TIMEOUT).await?;
```

Comment:
Wouldn't it make more sense to do this operation at the end of the split rather than at the beginning? Suppose something goes horribly wrong in the split failure.

Reply:
Either works, there are a couple of soft reasons for preferring doing it at the start:
- Disk space: generally it's more robust to do the operations that free space before the operations that consume it, so it's nice to detach the old secondaries before creating the new ones, rather than vice versa.
- At the end of the split, we remove the parent shards from in memory state.  If we were waiting for them to reconcile to detach secondary locations before doing that, then it would hold up that part of the process & if one of those reconciles failed, it would be a situation where we wouldn't want to bother aborting.

For the case where something goes horribly wrong to the point of taking the currently attached node offline, then this will indeed degrade availability for clients that could otherwise have used one of the secondary locations -- I think that's an okay tradeoff (the impact is limited to the tenant that was being split, and is temporary as it'll refill its cache with on-demand downloads).

All that said, looking at this again made me realize that on abort_tenant_shard_split we aren't calling schedule() on the parent shards, which is needed to reinstate their secondary locations after a split failure, so I've adding that and updating the failure test.

Covering secondaries exposed a safe-but-annoying behavior where we would refuse  to shard split when there was a secondary on an offline node, so there's a fix for that too, and a bonus commit that fixes preserving tenant configs in splits (the TODO happened to stick out to me while making these other changes)

Reply:
Mm, I hadn't considered disk space. I'm okay with this tradeoff.